projects/plugins/super-cache/wp-cache.php

```diff
@@ -3638,32 +3684,38 @@ function wpsc_reset_preload_settings() {
 
 function wpsc_cancel_preload() {
 	global $cache_path;
-	$next_preload = wp_next_scheduled( 'wp_cache_preload_hook' );
+	$next_preload      = wp_next_scheduled( 'wp_cache_preload_hook' );
+	$next_full_preload = wp_next_scheduled( 'wp_cache_full_preload_hook' );
+
+	if ( $next_preload || $next_full_preload ) {
+		wp_cache_debug( 'wpsc_cancel_preload: reset preload settings' );
+		wpsc_reset_preload_settings();
+	}
+
 	if ( $next_preload ) {
-		delete_transient( 'taxonomy_preload' );
 		wp_cache_debug( 'wpsc_cancel_preload: unscheduling wp_cache_preload_hook' );
-		wpsc_reset_preload_counter();
 		wp_unschedule_event( $next_preload, 'wp_cache_preload_hook' );
 	}
-	$next_preload = wp_next_scheduled( 'wp_cache_full_preload_hook' );
-	if ( $next_preload ) {
-		delete_transient( 'taxonomy_preload' );
-		wpsc_reset_preload_counter();
+	if ( $next_full_preload ) {
 		wp_cache_debug( 'wpsc_cancel_preload: unscheduling wp_cache_full_preload_hook' );
 		wp_unschedule_event( $next_preload, 'wp_cache_full_preload_hook' );
 	}
 	wp_cache_debug( 'wpsc_cancel_preload: creating stop_preload.txt' );
-	$fp = @fopen( $cache_path . "stop_preload.txt", 'w' );
-	@fclose( $fp );
+	$stop_preload_flag = $cache_path . 'stop_preload.txt';
+	if ( file_exists( $stop_preload_flag ) && time() - filemtime( $stop_preload_flag ) > 2 ) {
+		wp_cache_debug( 'wpsc_cancel_preload: force cancelled preload' );
+		wpsc_reset_preload_settings(); // reset the preload if we already tried more than 2 seconds ago.
```

Comment:
The 2 second wait is in case someone double tapped the "cancel" button. I'd rather the preload process stop the preload loop than the process serving the admin page.
But if the preload has got stuck or the user really wants to cancel immediately, then cancel it. That might have unforeseen consequences though - if the preload loop has started, then it will cache ten pages, and then schedule another loop.

I think it would be useful to make the preload loop more resilient. One way might be to check for the mutex file before scheduling the next loop. If it's not there it means the site owner cancelled the preload.

Reply:
That 2 second wait is now removed. When the user cancels the preload, all the settings are reset, and the "stop preload" file are created to make sure that the preload stops. I also added a check for that file to `wpsc_is_preload_active` and use that where I check for that "stop" file.